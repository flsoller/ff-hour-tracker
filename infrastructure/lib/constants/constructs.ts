export const USERS_SERVICE = {
  NAME: "hour-tracker-users-service",
  LAMBDA_INTEGRATION: "hour-tracker-users-service-lambda-integration",
};

export const API_GATEWAY = {
  NAME: "hour-tracker-api-gateway",
};

export const DEFAULT_AUTHORIZER = {
  NAME: "hour-tracker-default-authorizer-service",
  LAMBDA_AUTHORIZER: "hour-tracker-default-lambda-authorizer",
};

export const AUTHENTICATOR_SERVICE = {
  NAME: "hour-tracker-authentication-service",
  LAMBDA_INTEGRATION: "hour-tracker-authentication-service-lambda-integration",
};

export const ORGANIZATION_MANAGER = {
  NAME: "hour-tracker-organization-manager",
};
