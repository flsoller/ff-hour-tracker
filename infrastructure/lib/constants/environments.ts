export const ENVIRONMENT_PARAMS = {
  DATABASE_URL_KEY: "HOUR_TRACKER_DATABASE_URL",
  JWT_SECRET: "HOUR_TRACKER_JWT_SECRET",
  API_DOMAIN_NAME: "HOUR_TRACKER_API_DOMAIN_NAME",
  API_DOMAIN_CERT_ARN: "HOUR_TRACKER_API_DOMAIN_CERT_ARN",
};
