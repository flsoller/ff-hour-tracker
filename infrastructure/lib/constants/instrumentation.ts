export const DEFAULT_INSTRUMENTATION_CONFIG = {
  NEW_RELIC_EXTENSION_SEND_FUNCTION_LOGS: "true",
  NEW_RELIC_LAMBDA_EXTENSION_ENABLED: "true",
  NEW_RELIC_DATA_COLLECTION_TIMEOUT: "10s",
  NEW_RELIC_IGNORE_EXTENSION_CHECKS: "all",
  NEW_RELIC_NO_CONFIG_FILE: "true",
};
