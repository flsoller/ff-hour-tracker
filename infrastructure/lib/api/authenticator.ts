import {
  HttpApi,
  HttpMethod,
  HttpNoneAuthorizer,
} from "aws-cdk-lib/aws-apigatewayv2";
import { Duration, RemovalPolicy } from "aws-cdk-lib";
import * as ecr from "aws-cdk-lib/aws-ecr";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { HttpLambdaIntegration } from "aws-cdk-lib/aws-apigatewayv2-integrations";
import { Construct } from "constructs";
import { AUTHENTICATOR_SERVICE } from "../constants/constructs";
import { HOUR_TRACKER_ECR_REPO_NAMES } from "../constants/ecr";
import { LOGICAL_ID } from "../constants/logical-id";

interface AuthenticatorProps {
  hashOrVersion: string;
  dbConnectionString: string;
  apiGateway: HttpApi;
  jwtSecretString: string;
  newRelicAccountId: string;
  newRelicIngestLicense: string;
}

export class AuthenticationService extends Construct {
  constructor(scope: Construct, id: string, props: AuthenticatorProps) {
    super(scope, id);

    const ecrRepo = ecr.Repository.fromRepositoryName(
      this,
      `${id}-repo`,
      HOUR_TRACKER_ECR_REPO_NAMES.API_AUTHENTICATOR
    );

    const authenticatorService = new lambda.DockerImageFunction(
      this,
      AUTHENTICATOR_SERVICE.NAME,
      {
        code: lambda.DockerImageCode.fromEcr(ecrRepo, {
          tagOrDigest: props.hashOrVersion,
        }),
        architecture: lambda.Architecture.ARM_64,
        environment: {
          DATABASE_URL: props.dbConnectionString,
          JWT_SECRET: props.jwtSecretString,
          NEW_RELIC_LAMBDA_HANDLER: "apps/api/authenticator/index.handler",
          NEW_RELIC_ACCOUNT_ID: props.newRelicAccountId,
          NEW_RELIC_LICENSE_KEY: props.newRelicIngestLicense,
        },
        memorySize: 256,
        timeout: Duration.seconds(20),
        functionName: LOGICAL_ID.HOUR_TRACKER_API_AUTHENTICATOR,
      }
    );
    authenticatorService.applyRemovalPolicy(RemovalPolicy.DESTROY);

    const authenticatorServiceIntegration = new HttpLambdaIntegration(
      AUTHENTICATOR_SERVICE.LAMBDA_INTEGRATION,
      authenticatorService
    );

    props.apiGateway.addRoutes({
      integration: authenticatorServiceIntegration,
      path: "/auth/signin",
      methods: [HttpMethod.POST],
      authorizer: new HttpNoneAuthorizer(),
    });
  }
}
