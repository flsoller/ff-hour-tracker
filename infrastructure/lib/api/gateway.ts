import { Construct } from "constructs";
import {
  CorsHttpMethod,
  HttpApi,
  DomainName,
} from "aws-cdk-lib/aws-apigatewayv2";
import { Certificate } from "aws-cdk-lib/aws-certificatemanager";
import { HttpLambdaAuthorizer } from "aws-cdk-lib/aws-apigatewayv2-authorizers";
import { RemovalPolicy } from "aws-cdk-lib";
import { API_GATEWAY } from "../constants/constructs";
import { StringParameter } from "aws-cdk-lib/aws-ssm";
import { ENVIRONMENT_PARAMS } from "../constants/environments";

interface ApiGatewayProps {
  authService: HttpLambdaAuthorizer;
}

export class HourTrackerApiGateway extends Construct {
  public readonly httpApiGateway: HttpApi;

  constructor(scope: Construct, id: string, props: ApiGatewayProps) {
    super(scope, id);

    const customDomain = new DomainName(this, "DN", {
      domainName: StringParameter.valueForStringParameter(
        this,
        ENVIRONMENT_PARAMS.API_DOMAIN_NAME
      ),
      certificate: Certificate.fromCertificateArn(
        this,
        "CERT",
        StringParameter.valueForStringParameter(
          this,
          ENVIRONMENT_PARAMS.API_DOMAIN_CERT_ARN
        )
      ),
    });

    const gateway = new HttpApi(this, API_GATEWAY.NAME, {
      description: "REST API for hour tracker app",
      corsPreflight: {
        allowHeaders: ["Content-Type", "Authorization"],
        allowMethods: [
          CorsHttpMethod.OPTIONS,
          CorsHttpMethod.GET,
          CorsHttpMethod.POST,
          CorsHttpMethod.PUT,
          CorsHttpMethod.PATCH,
          CorsHttpMethod.DELETE,
        ],
        allowCredentials: true,
        allowOrigins: ["http://localhost:3000"],
      },
      createDefaultStage: false,
      defaultAuthorizer: props.authService,
      disableExecuteApiEndpoint: true,
    });
    gateway.applyRemovalPolicy(RemovalPolicy.DESTROY);
    gateway.addStage("live", {
      stageName: "live",
      throttle: {
        rateLimit: 1,
        burstLimit: 1,
      },
      autoDeploy: true,
      domainMapping: {
        domainName: customDomain,
      },
    });
    this.httpApiGateway = gateway;
  }
}
