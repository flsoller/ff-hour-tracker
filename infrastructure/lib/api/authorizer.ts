import { Construct } from "constructs";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as ecr from "aws-cdk-lib/aws-ecr";
import {
  HttpLambdaAuthorizer,
  HttpLambdaResponseType,
} from "aws-cdk-lib/aws-apigatewayv2-authorizers";
import { Duration, RemovalPolicy } from "aws-cdk-lib";
import { DEFAULT_AUTHORIZER } from "../constants/constructs";
import { HOUR_TRACKER_ECR_REPO_NAMES } from "../constants/ecr";
import { LOGICAL_ID } from "../constants/logical-id";
import { DEFAULT_INSTRUMENTATION_CONFIG } from "../constants/instrumentation";

interface AuthorizerProps {
  hashOrVersion: string;
  dbConnectionString: string;
  jwtSecretString: string;
  newRelicAccountId: string;
  newRelicIngestLicense: string;
}

export class AuthorizerService extends Construct {
  public readonly authorizer: HttpLambdaAuthorizer;

  constructor(scope: Construct, id: string, props: AuthorizerProps) {
    super(scope, id);

    const ecrRepo = ecr.Repository.fromRepositoryName(
      this,
      `${id}-repo`,
      HOUR_TRACKER_ECR_REPO_NAMES.API_AUTHORIZER
    );

    const authorizerService = new lambda.DockerImageFunction(
      this,
      DEFAULT_AUTHORIZER.NAME,
      {
        code: lambda.DockerImageCode.fromEcr(ecrRepo, {
          tagOrDigest: props.hashOrVersion,
        }),
        architecture: lambda.Architecture.ARM_64,
        environment: {
          DATABASE_URL: props.dbConnectionString,
          JWT_SECRET: props.jwtSecretString,
          NEW_RELIC_LAMBDA_HANDLER: "apps/api/authorizer/index.handler",
          NEW_RELIC_ACCOUNT_ID: props.newRelicAccountId,
          NEW_RELIC_LICENSE_KEY: props.newRelicIngestLicense,
          ...DEFAULT_INSTRUMENTATION_CONFIG,
        },
        memorySize: 256,
        timeout: Duration.seconds(20),
        functionName: LOGICAL_ID.HOUR_TRACKER_API_AUTHORIZER,
      }
    );
    authorizerService.applyRemovalPolicy(RemovalPolicy.DESTROY);

    this.authorizer = new HttpLambdaAuthorizer(
      DEFAULT_AUTHORIZER.LAMBDA_AUTHORIZER,
      authorizerService,
      {
        resultsCacheTtl: Duration.seconds(0),
        responseTypes: [HttpLambdaResponseType.SIMPLE],
      }
    );
  }
}
