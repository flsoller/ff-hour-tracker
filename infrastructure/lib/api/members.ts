import { Duration, RemovalPolicy } from "aws-cdk-lib";
import { HttpApi, HttpMethod } from "aws-cdk-lib/aws-apigatewayv2";
import { HttpLambdaIntegration } from "aws-cdk-lib/aws-apigatewayv2-integrations";
import * as ecr from "aws-cdk-lib/aws-ecr";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { Construct } from "constructs";
import { MEMBERS_SERVICE } from "../constants/constructs";
import { HOUR_TRACKER_ECR_REPO_NAMES } from "../constants/ecr";
import { DEFAULT_INSTRUMENTATION_CONFIG } from "../constants/instrumentation";
import { LOGICAL_ID } from "../constants/logical-id";

interface MembersProps {
  hashOrVersion: string;
  dbConnectionString: string;
  apiGateway: HttpApi;
  newRelicAccountId: string;
  newRelicIngestLicense: string;
}

export class MembersService extends Construct {
  constructor(scope: Construct, id: string, props: MembersProps) {
    super(scope, id);

    const ecrRepo = ecr.Repository.fromRepositoryName(
      this,
      `${id}-repo`,
      HOUR_TRACKER_ECR_REPO_NAMES.API_MEMBERS,
    );

    const membersService = new lambda.DockerImageFunction(
      this,
      MEMBERS_SERVICE.NAME,
      {
        code: lambda.DockerImageCode.fromEcr(ecrRepo, {
          tagOrDigest: props.hashOrVersion,
        }),
        architecture: lambda.Architecture.ARM_64,
        environment: {
          DATABASE_URL: props.dbConnectionString,
          NEW_RELIC_LAMBDA_HANDLER: "apps/api/members/index.handler",
          NEW_RELIC_ACCOUNT_ID: props.newRelicAccountId,
          NEW_RELIC_LICENSE_KEY: props.newRelicIngestLicense,
          ...DEFAULT_INSTRUMENTATION_CONFIG,
        },
        memorySize: 256,
        timeout: Duration.seconds(20),
        functionName: LOGICAL_ID.HOUR_TRACKER_API_MEMBERS,
      },
    );
    membersService.applyRemovalPolicy(RemovalPolicy.DESTROY);

    const membersServiceIntegration = new HttpLambdaIntegration(
      MEMBERS_SERVICE.LAMBDA_INTEGRATION,
      membersService,
    );

    props.apiGateway.addRoutes({
      integration: membersServiceIntegration,
      path: "/v1/members",
      methods: [HttpMethod.GET, HttpMethod.POST],
    });
  }
}
