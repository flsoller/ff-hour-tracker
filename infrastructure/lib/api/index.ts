import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import { AuthorizerService } from "./authorizer";
import { HourTrackerApiGateway } from "./gateway";
import { HOUR_TRACKER } from "../constants/stacks";
import * as ssm from "aws-cdk-lib/aws-ssm";
import { ENVIRONMENT_PARAMS } from "../constants/environments";
import { AuthenticationService } from "./authenticator";
import { MembersService } from "./members";

export class HourTrackerApi extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const hashOrVersion = new cdk.CfnParameter(this, "hashOrVersion", {
      description: "The version or hash used to reference the images",
      type: "String",
    }).valueAsString;

    const dbConnectionString = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.DATABASE_URL_KEY
    );

    const jwtSecretString = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.JWT_SECRET
    );

    const defaultAuthorizer = new AuthorizerService(
      this,
      HOUR_TRACKER.API_AUTHORIZER,
      {
        hashOrVersion,
        dbConnectionString,
        jwtSecretString,
      }
    );

    const apiGateway = new HourTrackerApiGateway(
      this,
      HOUR_TRACKER.API_GATEWAY,
      {
        authService: defaultAuthorizer.authorizer,
      }
    );

    new AuthenticationService(this, HOUR_TRACKER.API_AUTHENTICATOR, {
      apiGateway: apiGateway.httpApiGateway,
      dbConnectionString,
      hashOrVersion,
      jwtSecretString,
    });

    new MembersService(this, HOUR_TRACKER.API_MEMBERS, {
      apiGateway: apiGateway.httpApiGateway,
      dbConnectionString,
      hashOrVersion,
    });
  }
}
