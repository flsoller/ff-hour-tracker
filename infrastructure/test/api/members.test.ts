import { Template } from "aws-cdk-lib/assertions";
import * as cdk from "aws-cdk-lib";
import { HourTrackerApi } from "../../lib/api";
import { cdkResourceFinderWithTypeFilter } from "../helpers";
import { MEMBERS_SERVICE } from "../../lib/constants/constructs";

let app: cdk.App;
let apiStack: HourTrackerApi;
let template: Template;

beforeEach(() => {
  app = new cdk.App({
    context: { "@aws-cdk/core:newStyleStackSynthesis": false },
  });
  apiStack = new HourTrackerApi(app, "HourTrackerApi");
  template = Template.fromStack(apiStack);
});

describe("MembersServiceLambda", () => {
  it("should contain the correct Lambda properties", () => {
    const { DependsOn, ...attributesToCheck } = cdkResourceFinderWithTypeFilter(
      template,
      "AWS::Lambda::Function",
      MEMBERS_SERVICE.NAME
    );
    // Verify DependsOn is present but exclude it from snapshot
    expect(DependsOn).toBeDefined();
    expect(attributesToCheck).toMatchSnapshot();
  });
});
