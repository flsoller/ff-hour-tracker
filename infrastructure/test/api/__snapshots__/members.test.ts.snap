// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MembersServiceLambda should contain the correct Lambda properties 1`] = `
{
  "DeletionPolicy": "Delete",
  "Properties": {
    "Architectures": [
      "arm64",
    ],
    "Code": {
      "ImageUri": {
        "Fn::Join": [
          "",
          [
            {
              "Ref": "AWS::AccountId",
            },
            ".dkr.ecr.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/hour-tracker-members-service:",
            {
              "Ref": "hashOrVersion",
            },
          ],
        ],
      },
    },
    "Environment": {
      "Variables": {
        "DATABASE_URL": {
          "Ref": "SsmParameterValueHOURTRACKERDATABASEURLC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "NEW_RELIC_ACCOUNT_ID": {
          "Ref": "SsmParameterValueNEWRELICACCOUNTIDC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "NEW_RELIC_DATA_COLLECTION_TIMEOUT": "10s",
        "NEW_RELIC_EXTENSION_SEND_FUNCTION_LOGS": "true",
        "NEW_RELIC_IGNORE_EXTENSION_CHECKS": "all",
        "NEW_RELIC_LAMBDA_EXTENSION_ENABLED": "true",
        "NEW_RELIC_LAMBDA_HANDLER": "apps/api/members/index.handler",
        "NEW_RELIC_LICENSE_KEY": {
          "Ref": "SsmParameterValueNEWRELICLICENSEKEYC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "NEW_RELIC_NO_CONFIG_FILE": "true",
      },
    },
    "FunctionName": "HourTrackerAPI-Members",
    "MemorySize": 256,
    "PackageType": "Image",
    "Role": {
      "Fn::GetAtt": [
        "hourtrackerapimembershourtrackermembersserviceServiceRole72DF1FAF",
        "Arn",
      ],
    },
    "Timeout": 20,
  },
  "Type": "AWS::Lambda::Function",
  "UpdateReplacePolicy": "Delete",
}
`;
